components/builder-api/src/server/resources/origins.rs

```diff
@@ -516,46 +516,61 @@ fn upload_origin_key(req: HttpRequest,
                      -> HttpResponse {
     let (origin, revision) = path.into_inner();
 
-    let account_id =
-        match authorize_session(&req, Some(&origin), Some(OriginMemberRole::Administrator)) {
-            Ok(session) => session.get_id(),
-            Err(err) => return err.into(),
-        };
+    // Since this route allows users to upload keys, we verify their membership
+    // before we determine if the key they're using actually exists
+    if let Err(err) = authorize_session(&req, Some(&origin), Some(OriginMemberRole::Member)) {
+        return err.into();
+    }
 
     let conn = match state.db.get_conn().map_err(Error::DbError) {
         Ok(conn_ref) => conn_ref,
         Err(err) => return err.into(),
     };
 
-    match parse_key_str(&body) {
-        Ok((PairType::Public, ..)) => {
-            debug!("Received a valid public key");
-        }
-        Ok(_) => {
-            debug!("Received a secret key instead of a public key");
-            return HttpResponse::new(StatusCode::UNPROCESSABLE_ENTITY);
-        }
-        Err(e) => {
-            debug!("Invalid public key content: {}", e);
-            return HttpResponse::new(StatusCode::UNPROCESSABLE_ENTITY);
-        }
-    }
+    match OriginPublicSigningKey::get(&origin, &revision, &conn) {
+        Ok(_) => HttpResponse::new(StatusCode::CONFLICT),
+        Err(_) => {
```

Comment:
Minor - We could get rid of this `match` by having:

```
if OriginPublicSigningKey::get(&origin, &revision, &conn).is_ok() {
    return HttpResponse::new(StatusCode::CONFLICT);
}
```

and then the `Err` case of the `match` would just be the rest of the function.